pyproject.toml

```
@@ -87,9 +85,9 @@ docs = [
 ]
 optional = [
     'SimpleITK',
-    'astropy>=5.0',
-    'cloudpickle>=0.2.1',  # necessary to provide the 'processes' scheduler for dask
-    'dask[array]>=2021.1.0,!=2024.8.0',
+    'astropy>=6.0',
+    'cloudpickle>=3.0',  # necessary to provide the 'processes' scheduler for dask
```

Local testing with `uv pip` shows that `cloudpickle>=1.1.1` should also work?

```suggestion
    'cloudpickle>=1.1.1',  # necessary to provide the 'processes' scheduler for dask
```

---

Cloudpickle 1.1.1 was released May 15 2019. Why would we need to support such an old version?

This recommendation exactly follows SPEC0, so let's move ahead.

---

Kind of disagree here. I'm all for dropping old versions according to SPEC0 if we gain something, e.g., remove workarounds for old API's, etc. But there seems to be no such advantage here and no disadvantage to keeping the old version around. So this feels unnecessary. Am I missing something?

On the other hand we restrict the space of dependency resolution for users with our package even though cloudpickle==1.1.1 would work perfectly fine.

I fuzzily remember a blog post or a discussion on this topic? Makes it especially hard to find valid solutions with tools that do "correct" dependency resolution? But that might have been discussing the whole `^=` thing from poetry.

---

It's a matter of giving ourselves the license of not having to worry about supporting older packages. This license is exactly what SPEC0 is. We could pin an older version, but when Jarrod upped dependencies, he tested with a *valid newer version according to SPEC0*. You re-tested with an older version (why?), but now Jarrod would have to go and verify that it works, modify his PR, etc. etc. It all costs. We have *so little* person-power. We haven't built significant image processing features in years—each release is just a string of refactorings and fixes and maintenance changes. We are behind on our grant outcomes. We get stuck on each release. It is *imperative* that we do *nothing* to slow down developers who are *operating within our chosen rules*.

If the core team disagrees, in that they'd prefer to slow things down, and make it hard to get the project goals—that we agreed on jointly—knocked out, on a project that is moving forward at glacial speed already, I would be most disappointed.

Our users are important. They're who we do this for. Do they benefit from getting support for an old version of cloudpickle? Maybe—a tiny, tiny bit? Do they benefit from us wasting time and not getting out new features? Absolutely not. We, as a team, need to take an urgent look at what our priorities are going to be.

---

I'm sorry, I don't want to cause more frustration with this. I've been feeling the frustration around lacking person-power too. I'll try to reflect on your comment a bit more and see if I can come up with something that moves this forward in general. It feels similar to past conflicts we've had.

So I'll only answer this specific question for now.

> You re-tested with an older version (why?), but now Jarrod would have to go and verify that it works, modify his PR, etc. etc.

When reviewing I try to understand the reason for changes. I couldn't for this PR because there was no explanation given in the description of PR or commit. So I tested locally with an older version to maybe find the reason myself. I couldn't so I raised the question. 

Note that I didn't interpret SPEC0 to cover all dependencies, including cloudpickle, rather only the core dependencies listed on it.

---

Could you clarify that last sentence? The SPEC is meant to give developers license to drop *any* library dependency of a certain age.

Also noted that it would be helpful to make notes in commit messages (or comments) on why certain updates are needed.

---

> Could you clarify that last sentence? The SPEC is meant to give developers license to drop any library dependency of a certain age.

The SPEC seems to refer "dependencies in question are core packages" so I might have gotten the wrong impression from that. That said, the motivation behind the SPEC applies to every kind of dependency.

---

You are right; I didn't notice that before. @jarrodmillman should we remove the word "core" in that sentence?